File: 25-reverse-nodes-in-k-group/25-reverse-nodes-in-k-group.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def reverseKGroup(self, head: Optional[ListNode], k: int) -> Optional[ListNode]:
        prev, current = None, head
        for _ in range (k):
            if current is None:
                return head
            current = current.next
        current = head
        for _ in range (k):
            nxt = current.next
            current.next = prev
            prev = current
            current = nxt
        head.next = self.reverseKGroup(current, k)
        return prev

File: 25-reverse-nodes-in-k-group/25-reverse-nodes-in-k-group.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode reverseKGroup(ListNode head, int k) {
        ListNode prev = new ListNode(-1, null);
        ListNode current = head;
        for(int i = 0; i < k; ++i) {
            if(current == null) return head;
            current = current.next;
        }
        current = head;
        for(int i = 0; i < k; ++i) {
            ListNode next = current.next;
            current.next = prev;
            prev = current;
            current = next;
        }
        head.next = reverseKGroup(current, k);
        return prev;
    }
}